pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -522,63 +501,129 @@ func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet,
 		}
 
-		if podCount != groupTotalCount {
-			errMsg := fmt.Sprintf("'%s' group total count is different from the actual number of pods in the cluster", p.groupName())
-			r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", errMsg)
-			return nil, jobframework.UnretryableError(errMsg)
+		tc, err := strconv.Atoi(podInGroup.GetAnnotations()[GroupTotalCountAnnotation])
+		if err != nil {
+			return fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
 		}
+		if tc != groupTotalCount {
+			return jobframework.UnretryableError(fmt.Sprintf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.pod.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc))
+		}
+	}
 
-		podSets, err = p.constructGroupPodSets(p.list)
-		if err != nil {
-			if jobframework.IsUnretryableError(err) {
-				r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", err.Error())
+	return nil
+}
+
+// activePods returns a slice of non-failed pods in the group
+func (p *Pod) activePods() []corev1.Pod {
+	// Filter all failed pods
+	activePods := make([]corev1.Pod, 0, len(p.list.Items))
+	for _, pod := range p.list.Items {
+		if pod.Status.Phase != corev1.PodFailed {
+			activePods = append(activePods, pod)
+		}
+	}
+
+	return activePods
+}
+
+// cleanupExcessPods will delete and finalize pods created last if the number of
+// activePods is greater than the totalCount value.
+func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, totalCount int, activePods []corev1.Pod) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	extraPodsCount := len(activePods) - totalCount
+
+	if extraPodsCount <= 0 {
+		return nil
+	}
+
+	// Sort active pods by creation timestamp
+	sort.SliceStable(activePods, func(i, j int) bool {
+		return activePods[i].ObjectMeta.CreationTimestamp.Before(&activePods[j].ObjectMeta.CreationTimestamp)
+	})
+
+	// Extract all the latest created extra pods
+	extraPods := activePods[len(activePods)-extraPodsCount:]
+
+	// Finalize and delete the active pods created last
+	for _, extraPod := range extraPods {
+		if controllerutil.RemoveFinalizer(&extraPod, PodFinalizer) {
+			log.V(3).Info("Finalizing excess pod in group", "ExcessPod.Name", extraPod.Name)
+			if err := c.Update(ctx, &extraPod); err != nil {
+				return err
+			}
+		}
+		if extraPod.ObjectMeta.DeletionTimestamp.IsZero() {
+			log.V(3).Info("Deleting excess pod in group", "ExcessPod.Name", extraPod.Name)
+			if err := c.Delete(ctx, &extraPod); err != nil {
+				return err
 			}
-			return nil, err
 		}
 
-		if len(podSets) > 8 {
-			return nil, jobframework.UnretryableError(errMsgIncorrectGroupRoleCount)
+		// Remove excess pod from the group list
+		for i := range p.list.Items {
+			if p.list.Items[i].Name == extraPod.Name && p.list.Items[i].Namespace == extraPod.Namespace {
+				p.list.Items = append(p.list.Items[:i], p.list.Items[i+1:]...)
+			}
```

Comment:
This is an expensive way to remove items from a list, as it will do allocations for every deleted item.

Better generate a new list, like so:

```golang
newList = make([]Pods, 0, len(p.list.Items)-len(extra))
for _, pod := range p.list.Items {
  if (pod not found in extra) {
    newList = append(newList, p)
  }
}
p.list.Items = newList
```